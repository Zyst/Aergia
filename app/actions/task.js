// @flow
import type { actionType } from "../reducers/task";

export const ADD_TASK = "ADD_TASK";
export const COMPLETE_TASK = "COMPLETE_TASK";
export const REMOVE_TASK = "REMOVE_TASK";
export const ACTIVATE_TASK = "ACTIVATE_TASK";
export const DEACTIVATE_TASK = "DEACTIVATE_TASK";

export function addTask(name: string, totalPomodoros: number): actionType {
  return {
    type: ADD_TASK,
    name,
    payload: {
      totalPomodoros
    }
  };
}

export function completeTask(name: string): actionType {
  return {
    type: COMPLETE_TASK,
    name,
    payload: {}
  };
}

export function removeTask(name: string): actionType {
  return {
    type: REMOVE_TASK,
    name,
    payload: {}
  };
}

export function activateTask(name: string): actionType {
  return {
    type: ACTIVATE_TASK,
    name,
    payload: {}
  };
}

export function deactivateTask(name: string): actionType {
  return {
    type: DEACTIVATE_TASK,
    name,
    payload: {}
  };
}
