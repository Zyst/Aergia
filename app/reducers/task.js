// @flow
import {
  ADD_TASK,
  COMPLETE_TASK,
  REMOVE_TASK,
  ACTIVATE_TASK,
  DEACTIVATE_TASK,
  POMODORO_DONE
} from "../actions/task";

export type taskStateType = Array<{
  name?: string,
  currentProgress?: number,
  totalPomodoros?: number,
  stopped?: number,
  active?: boolean
}>;

export type actionType = {
  type: string,
  name?: string,
  payload?: {
    currentProgress?: number,
    totalPomodoros?: number,
    stopped?: number,
    active?: boolean
  }
};

export default function tasks(
  state: taskStateType = [
    {
      name: "Testing",
      currentProgress: 0,
      totalPomodoros: 1,
      stopped: 0,
      active: false
    },
    {
      name: "Active test",
      currentProgress: 1,
      totalPomodoros: 3,
      stopped: 1,
      active: true
    }
  ],
  action: actionType
): taskStateType {
  switch (action.type) {
    case ADD_TASK:
      if (
        // If a task with that name doesn't exist already
        state.filter(task => {
          const taskName = task.name || "";
          const actionName = action.name || "";

          return (
            taskName.toLocaleLowerCase() === actionName.toLocaleLowerCase()
          );
        }).length === 0 &&
        action.payload
      ) {
        return [
          ...state,
          {
            name: action.name,
            currentProgress: 0,
            totalPomodoros: action.payload.totalPomodoros,
            stopped: 0,
            active: false
          }
        ];
      }
      return state;

    // Although this is currently the same as remove, that will change later
    case COMPLETE_TASK:
    case REMOVE_TASK:
      return state.filter(task => task.name !== action.name);

    case ACTIVATE_TASK:
      return state.map(task => {
        if (task.name === action.name) {
          return {
            ...task,
            active: true
          };
        }
        return task;
      });

    case DEACTIVATE_TASK:
      return state.map(task => {
        if (task.name === action.name) {
          return {
            ...task,
            // We increment stopped if the task was active
            stopped: task.active ? task.stopped + 1 : task.stopped,
            active: false
          };
        }
        return task;
      });

    case POMODORO_DONE:
      return state.map(task => {
        if (task.active) {
          return {
            ...task,
            currentProgress: task.currentProgress + 1
          };
        }
        return task;
      });

    default:
      return state;
  }
}
