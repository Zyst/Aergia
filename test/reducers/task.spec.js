import task from "../../app/reducers/task";
import {} from "../../app/actions/task";

describe("reducers", () => {
  describe("task", () => {
    it("should be defined", () => {
      expect(task).toBeDefined();
    });

    it("should handle initial state", () => {
      expect(task(undefined, [])).toMatchSnapshot();
    });
  });
});
