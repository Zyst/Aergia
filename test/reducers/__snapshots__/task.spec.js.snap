// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`reducers task should handle initial state 1`] = `Array []`;
